dask_geomodeling/tests/test_raster_parallelize.py

```diff
@@ -0,0 +1,183 @@
+    )
+    check_sources_and_requests(s_r, expected_tiles)
+
+
+@pytest.mark.parametrize(
+    "cellsize,expected_tiles",
+    [
+        ((1, 1), [(3.0, 3.0, 15.0, 15.0)]),
+        ((2, 2), [(2.0, 2.0, 16.0, 16.0)]),
+        ((3, 3), [(3.0, 3.0, 15.0, 15.0)]),
+        ((4, 4), [(0.0, 0.0, 16.0, 16.0)]),
+        ((2, 3), [(2.0, 3.0, 16.0, 15.0)]),
+        ((1, 4), [(3.0, 0.0, 15.0, 16.0)]),
+    ],
+)
+def test_tiling_cellsize(empty_source, cellsize, expected_tiles):
+    block = raster.RasterTiler(empty_source, 24, "EPSG:28992")
+    s_r = block.get_sources_and_requests(
+        mode="vals",
+        bbox=(3.0, 3.0, 15.0, 15.0),
+        width=int(12 / cellsize[0]),
+        height=int(12 / cellsize[1]),
+        projection="EPSG:28992",
+    )
+    check_sources_and_requests(s_r, expected_tiles, cellsize)
+
+
+@pytest.mark.parametrize(
+    "corner,expected_tiles",
+    [
+        ((0, 0), [(2.0, 2.0, 5.0, 5.0)]),
+        ((0.1, 0), [(1.1, 2.0, 5.1, 5.0)]),
+        ((0, 0.1), [(2.0, 1.1, 5.0, 5.1)]),
+        ((0.1, 0.1), [(1.1, 1.1, 5.1, 5.1)]),
+    ],
+)
+def test_tiling_corner(empty_source, corner, expected_tiles):
+    block = raster.RasterTiler(empty_source, 7, "EPSG:28992", corner)
+    s_r = block.get_sources_and_requests(
+        mode="vals",
+        bbox=(2.0, 2.0, 5.0, 5.0),
+        width=3,
+        height=3,
+        projection="EPSG:28992",
+    )
+    check_sources_and_requests(s_r, expected_tiles)
+
+
+def test_cell_tile_mismatch(empty_source):
+    block = raster.RasterTiler(empty_source, 7, "EPSG:28992")
+    s_r = block.get_sources_and_requests(
+        mode="vals",
+        bbox=(0.0, 0.0, 6.0, 6.0),
+        width=2,  # cellsize_x = 3 will be adjusted with round(7 / 3) to 3.5
+        height=3,  # cellsize_y = 2 will be adjusted with round(7 / 2) to 1.75
+        projection="EPSG:28992",
+    )
+    s_r = list(s_r)[1:]
+    assert len(s_r) == 1
+    assert s_r[0][1]["width"] == 2
+    assert s_r[0][1]["height"] == 4
+
+
+@pytest.mark.parametrize(
+    "bbox_offset",
+    [
+        (0, -5, 5, 0),  # covers precisely the 5x5 meters of data in the source
+        (0, -5, 6, 0),
+        (0, -6, 5, 0),
+        (-1, -5, 5, 0),
+        (0, -5, 5, 1),
+    ],
+)
+def test_tiling_process(source, bbox_offset):
+    # piece back together tiles with nodata at the negative y edge
+    block = raster.RasterTiler(source, 2, "EPSG:28992")
+    request = dict(
+        mode="vals",
+        bbox=(
+            source.pixel_origin[0] + bbox_offset[0],
+            source.pixel_origin[1] + bbox_offset[1],
+            source.pixel_origin[0] + bbox_offset[2],
+            source.pixel_origin[1] + bbox_offset[3],
+        ),
+        width=(bbox_offset[2] - bbox_offset[0]) * 2,  # 0.5 m resolution
+        height=(bbox_offset[3] - bbox_offset[1]) * 2,  # 0.5 m resolution
+        projection="EPSG:28992",
+        start=source.period[0],
+    )
+    actual = block.get_data(**request)
+    expected = source.get_data(**request)
```

Comment:
Just out of interest, this would be a test where we could notice the effects of utils.zoom_raster() if the bbox does not align nicely with cell edges, don't you think?

Reply:
Yes, I think so, but this test should be on the MemorySource itself

Reply:
Ok!